### PR TITLE
Hook up Slurm API to DWS API.

### PR DESCRIPTION
**Pull request by roehrich-hpe**
Remove burst_buffer.lua placeholder, move DWS library to that name.

Add helpers to glue the two APIs together, and a new batch of tests.

Update DWS to get latest changes there, including new spellings for the state names.  The state names are now CamelCase.

In KIND, use kubeadm to label worker node.

Signed-off-by: Dean Roehrich <dean.roehrich@hpe.com>

### CODE REVIEW

File: src/burst_buffer/burst_buffer.lua [hunk 1]
```diff
@@ -1,39 +1,730 @@
...
+DWS = {
+	-- The name of the Workflow resource.
+	name = "",
+
+	-- The YAML text of the Workflow resource.
+	yaml = "",
+
+--	-- Used by the tests when stubbing io.popen.  These values will be
+--	-- returned by DWS:io_popen().
+--	mock_io_popen = {
```
Comment (roehrich-hpe) [inline src/burst_buffer/burst_buffer.lua:102]: cruft to cleanup.

File: src/burst_buffer/burst_buffer.lua [hunk 2]
```diff
@@ -1,39 +1,722 @@
...
+--and there will be no need for cleanup.
+--
+--If this function returns an error, the job is rejected and the second return
+--value (if given) is printed where salloc, sbatch, or srun was called.
+--]]
+function slurm_bb_job_process(job_script)
+	slurm.log_info("%s: slurm_bb_job_process(). job_script=%s",
+		lua_script_name, job_script)
+
+	-- Note: In this version of Slurm we do not have the job ID in this
```
Comment (nathandotleeathpe) [inline src/burst_buffer/burst_buffer.lua:408]: Specify which version of Slurm "this version" is. You should also be specific about which version of Slurm introduces the job id in slurm_bb_job_process.

File: src/burst_buffer/burst_buffer.lua [hunk 3]
```diff
@@ -1,39 +1,722 @@
...
+	-- and with placeholder values for the user ID and group ID.  We will
+	-- submit it, report on whether it was good, and then delete it.  The
+	-- slurm_bb_setup() stage will have to re-create it using the job ID to
+	-- name it.
+
+	local workflow_name = "temp-" .. math.random(10000)
+	local workflow = DWS(workflow_name)
+	local done, err = make_workflow(workflow, job_script, 1, 1, 1)
+	if done == false then
+		slurm.log_error("%s: slurm_bb_job_process(). %s", lua_script_name, err)
```
Comment (nathandotleeathpe) [inline src/burst_buffer/burst_buffer.lua:423]: Do we get the temp workflow name here?

File: src/burst_buffer/burst_buffer.lua [hunk 4]
```diff
@@ -1,39 +1,722 @@
...
+\"pools\":\
+  [\
+    { \"id\":\"pool1\", \"quantity\":1000, \"granularity\":1024 },\
+    { \"id\":\"pool2\", \"quantity\":5, \"granularity\":2 },\
+    { \"id\":\"pool3\", \"quantity\":4, \"granularity\":1 },\
+    { \"id\":\"pool4\", \"quantity\":25000, \"granularity\":1 }\
+  ]\
+}"
+
+	return slurm.SUCCESS, pools
```
Comment (nathandotleeathpe) [inline src/burst_buffer/burst_buffer.lua:487]: Leave a comment here that we are ignoring the `slurm_bb_pools` function. We can also comment out the log message and replace the pool list with a placeholder pool.

File: src/burst_buffer/burst_buffer.lua [hunk 5]
```diff
@@ -1,39 +1,722 @@
...
+	if done == false then
+		slurm.log_error("%s: slurm_bb_setup(). %s", lua_script_name, err)
+		return slurm.ERROR, err
+	end
+
+	-- Wait for proposal state to complete, or pick up any error that may
+	-- be waiting in the Workflow.
+	done, err = workflow:wait_for_status_complete(60)
+	if done == false then
+		slurm.log_error("%s: slurm_bb_setup(). %s", lua_script_name, err)
```
Comment (nathandotleeathpe) [inline src/burst_buffer/burst_buffer.lua:546]: We need more information here about what we are waiting for. An error here would just say "exceeded max wait time" and would not indicate if we timed out during the creation or setup step.

File: src/burst_buffer/burst_buffer.lua [hunk 6]
```diff
@@ -1,39 +1,722 @@
...
+
+	local done, err = workflow:set_desired_state("Setup")
+	if done == err then
+		slurm.log_error("%s: slurm_bb_setup(). %s", lua_script_name, err)
+		return done, err
+	end
+
+	done, err = workflow:wait_for_status_complete(60)
+	if done == err then
+		slurm.log_error("%s: slurm_bb_setup(). %s", lua_script_name, err)
```
Comment (nathandotleeathpe) [inline src/burst_buffer/burst_buffer.lua:558]: See above comment.
> We need more information here about what we are waiting for. An error here would just say "exceeded max wait time" and would not indicate if we timed out during the creation or setup step.

File: src/burst_buffer/burst_buffer.lua [hunk 7]
```diff
@@ -1,39 +1,722 @@
...
+--pending.
+--
+--This function is only called if pools are specified and the job requested a
+--pool. This function may return a number (surrounded by quotes to make it a
+--string) as the second return value. If it does, the job's usage of the pool
+--will be changed to this number. A commented out example is given.
+--]]
+function slurm_bb_real_size(job_id)
+	slurm.log_info("%s: slurm_bb_real_size(). job id:%s",
+		lua_script_name, job_id)
```
Comment (nathandotleeathpe) [inline src/burst_buffer/burst_buffer.lua:600]: If this is a no-op, do we need to leave the log message?

**Review by nathandotleeathpe — APPROVED** [review]

**Pull request merged**